### PR TITLE
GitHub Filesystem block

### PR DESCRIPTION
**Pull request by cicdw**
### Summary

This PR introduces the minimum necessary interface to support GitHub as a remote storage option when configuring deployments; it does not yet wire up that support to the deployments interfaces (that will be in a future PR).

### Steps Taken to QA Changes

Tested locally for now, with and without references.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- No tests or issue needed
- [ ] A short code fix
	- Please reference the related issue by including "closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a bug report issue 
	- Please include tests. One-line fixes without tests will not be accepted unless it's related to the documentation only.
- [x] A new feature implementation
	- Please reference the related issue by including "closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a feature enhancement issue 
	- Please include tests
    - Please make sure that your QA steps are both thorough and easy to reproduce by somebody with limited knowledge of the feature that you are submitting

**Happy engineering!**

### CODE REVIEW

**Comment by desertaxle** [comment]
I think we should create interface classes for `get_directory` and `put_directory` since those have become the de factor standard for what's needed for a block to work for deployment flow storage. Have defined block schema capabilities corresponding to those methods will help us filter and verify which blocks can be used for deployment storage.

I think `read_path` and `write_path` will be useful for results storage and are worth keeping around for that purpose, but someone with a more informed idea of what remote result storage will look like should weigh in on that.

**Review by anna-geller — APPROVED** [review]
LGTM

File: tests/test_filesystems.py [hunk 1]
```diff
@@ -87,3 +87,12 @@ async def test_resolve_path(self):
         assert fs._resolve_path(base) == base + "/"
         assert fs._resolve_path(f"{base}/subdir") == f"{base}/subdir"
         assert fs._resolve_path("subdirectory") == f"{base}/subdirectory"
+
+
+class TestGitHub:
+    async def test_subprocess_errors_are_surfaced(self):
```
Comment (bunchesofdonald) [inline tests/test_filesystems.py:93]: `get_directory` has quite a bit of branching, might be nice to cover some of those other paths.

**Review by bunchesofdonald — APPROVED** [review]
I like that we're starting with just a read-only version, writing is a huge 'can-of-worms'.

File: src/prefect/filesystems.py [hunk 2]
```diff
@@ -723,3 +725,48 @@ async def read_path(self, path: str) -> bytes:
 
     async def write_path(self, path: str, content: bytes) -> str:
         return await self.filesystem.write_path(path=path, content=content)
+
+
+class GitHub(ReadableDeploymentStorage):
```
Comment (desertaxle) [inline src/prefect/filesystems.py:730]: Should we name this more broadly to something like `GitRepository`? There's nothing in this block that is specific to GitHub.

Comment (anna-geller) [reply src/prefect/filesystems.py:730]: that's technically true, but this way we would name S3 as S3Path -- here I think we name it all based on system/service for storing files and directories; GitHub is an outlier here anyway, so this totally could be a GitRepository

Comment (cicdw) [reply src/prefect/filesystems.py:730]: @desertaxle thinking ahead, the next change will almost certainly be to expose authentication to this block which I believe will be GitHub specific

File: tests/test_filesystems.py (hunk 1 above)
Comment (cicdw) [reply tests/test_filesystems.py:95]: @bunchesofdonald makes sense, I added a few critical path tests

**Pull request merged**